### ISSUE TITLE
specifying shell/$SHELL path on windows

### ISSUE THREAD

**Issue by CervEdin** [neovim/neovim#16646]
### Neovim version (nvim -v)

0.6.0

### Vim (not Nvim) behaves the same?

no 8.2.3441

### Operating system/version

Windows 10/19043.1348

### Terminal name/version

Windows terminal 1.11.2921.0

### $TERM environment variable

none in powershell, xterm-256color in git-bash

### Installation

scoop

### How to reproduce the issue

1 Run `powershell` or `mingw-bash.exe` in windows terminal or the stand-alone `neovim` app.
2 Try and set `shell` to a path.

### Expected behavior

According to the [shell documentation](https://neovim.io/doc/user/options.html#'shell'), one can specify the shell pathname like this:

```
Note the backslash before each quote (to avoid starting a comment) and 
	each space (to avoid ending the option value), so better use |:let-&| 
	like this:
		:let &shell='"C:\Program Files\unix\sh.exe" -f'
```
It should be noted that I'm trying to set `C:\Program Files\Git\bin\bash.exe -i -l` as shell but since it has spaces I'd first expect to be able to set the more basic `C:\Windows\System32\cmd.exe` first.

Ideally, this would work just like in `vim` where my `SHELL` is `/usr/bin/bash` and everything works.
If that's not possible it'd be great if there's a way to manually set my shell to `C:\Program Files\Git\usr\bin\bash`.

### Actual behavior

# Setting the environment SHELL variable to a path
In `powershell`
```ps
$Env:SHELL = "C:\Windows\System32\cmd.exe"
```
In `git-bash`
```sh
env SHELL='C:\Windows\System32\cmd.exe' nvim
```
## result
```vim
:echo $SHELL
C:\Windows\System32\cmd.exe
:ter
E475: Invalid value for argument cmd: 'C:WindowsSystem32cmd.exe' is not executable
```
# Using let-& (wo environment variable)
```vim
:let &shell='"C:\Windows\System32\cmd.exe"'
:ter
E475: Invalid value for argument cmd: 'C:WindowsSystem32cmd.exe' is not executable
```
# Using let-& forward slashes
```vim
:let &shell='"c:/Windows/System32/cmd.exe"'
A subdirectory or file .exe already exists.
[Process exited 1]
term://~//28128:c:/Windows/System32/cmd.exe
```
# Using a path with spaces
```vim
:let &shell='"C:\Program Files\Git\bin\bash"'
E475: Invalid value for argument cmd: 'C:Program FilesGitusr^Hin^Hash.exe' is not executable
```
# Specifying bash in nvim in `git-bash`
```
:ter bash
/usr/bin/bash: /s: No such file or directory
[Process exited 127]
```

**Comment by zeertzjq** [comment]
Possible solution:
```diff
diff --git a/src/nvim/ex_docmd.c b/src/nvim/ex_docmd.c
index 9f0f8d93a..3e78a5855 100644
--- a/src/nvim/ex_docmd.c
+++ b/src/nvim/ex_docmd.c
@@ -9720,7 +9720,9 @@ static void ex_terminal(exarg_T *eap)
     char shell_argv[512] = { 0 };
 
     while (*p != NULL) {
-      snprintf(tempstring, sizeof(tempstring), ",\"%s\"", *p);
+      char *escaped = (char *)vim_strsave_escaped((char_u *)*p, (char_u *)"\"\\");
+      snprintf(tempstring, sizeof(tempstring), ",\"%s\"", escaped);
+      xfree(escaped);
       xstrlcat(shell_argv, tempstring, sizeof(shell_argv));
       p++;
     }
```

**Comment by zeertzjq** [comment]
> A subdirectory or file .exe already exists.

This may indicate a different problem.

**Comment by CervEdin** [comment]
@zeertzjq yes I think so. Though `C:/Windows/System32/cmd.exe` is a valid but uncommon path.

I tried it again on a different machine. First I just get.
```
[Process exited 0]
```
then its
```
A subdirectory or file .exe already exists
[Process exited 1]
```

**Comment by moeshin** [comment]
> ## Specifying bash in nvim in git-bash
> ```text
> :ter bash
> /usr/bin/bash: /s: No such file or directory
> [Process exited 127]
> ```

https://github.com/neovim/neovim/issues/16957

```vim
echo &shell
" D:/msys64/usr/bin/bash

echo &shellcmdflag
" /s /c

ter bash
" /usr/bin/bash: /s: No such file or directory
" [Process exited 127]

let &shellcmdflag = '-c'
ter bash
" work
```

**Comment by moeshin** [comment]
```vim
let &shell = '"C:\\Program Files\\Git\\bin\\bash.exe"'
```

**Comment by 3N4N** [comment]
I don't know why the doc uses only one backslash.  I distinctly remember reading (I'm sure somewhere in the helpdocs even though I can't find it now), that you need to escape the backslash twice (so, total 4 backslashes).  I have this in my config:
```
let &shell = "C:\\\\Windows\\\\System32\\\\cmd.exe"
" which can be easily converted to what's below
let &shell = '"C:\\\\msys64\\\\usr\\\\bin\\\\bash.exe" --help'
```

@zeertzjq comment?

**Comment by zeertzjq** [comment]
Backslash don't have a special meaning in Vim script single-quoted strings.

**Comment by 3N4N** [comment]
Got it.  Thanks.  I realize now that what's said in the helpdoc is the expected behavior.

**Comment by OuyangIPC** [comment]
> ```viml
> let &shell = '"C:/Program Files/Git/bin/bash.exe"'
> ter " ok
> ```
> 
> ```viml
> ter C://Windows//System32//cmd.exe " ok cmd and bash
> ter "C://Program Files//Git//bin//bash.exe" " ok only cmd
> ter "'C:/Program Files/Git/bin/bash.exe'" " ok only bash and let &shellcmdflag = '-c'
> ```

May I ask a question. "let &shell=" how to write in init.lua?  thank you a lot.

**Comment by zeertzjq** [comment]
`vim.o.shell = `

**Comment by OuyangIPC** [comment]
> `vim.o.shell = `

@zeertzjq OK，thank you a lot.
